c7n/filters/iamaccess.py

```diff
@@ -343,3 +343,116 @@ def __call__(self, r):
         if violations:
             r[self.annotation_key] = violations
             return True
+
+
+class PolicyStatementFilter(Filter):
+    """  Check a resource for a set of policy statements.
+    """
+    POLICY_ELEMENTS = [
+        'Sid', 'Effect', 'Principal', 'NotPrincipal', 'Action',
+        'NotAction', 'Resource', 'NotResource', 'Condition']
+    schema = type_schema(
+        'has-statement',
+        statement_ids={'type': 'array', 'items': {'type': 'string'}},
+        statements={
+            'type': 'array',
+            'items': {
+                'type': 'object',
+                'properties': {
+                    'Sid': {'type': 'string'},
+                    'Effect': {'type': 'string', 'enum': ['Allow', 'Deny']},
+                    'Principal': {'anyOf': [
+                        {'type': 'string'},
+                        {'type': 'object'}, {'type': 'array'}]},
+                    'NotPrincipal': {
+                        'anyOf': [{'type': 'object'}, {'type': 'array'}]},
+                    'Action': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'NotAction': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Resource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'NotResource': {
+                        'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                    'Condition': {'type': 'object'},
+                    'PartialMatch': {
+                        'anyOf': [
+                            {'type': 'string', "enum": POLICY_ELEMENTS},
+                            {'type': 'array', 'items': [
+                                {"type": "string", "enum": POLICY_ELEMENTS}]}]
+                    }
+                },
+                'required': ['Effect']
+            }
+        })
+
+    def process(self, resources, event=None):
+        return list(filter(None, map(self.process_resource, resources)))
+
+    def get_policy(self, r):
+        return r.get('Policy')
```

Comment:
rather than reimplementing this, it might be better to just compose with cross-account filter to get the resource policy, that would also obviate alot of the per resource tests and recordings in this pr.

Reply:
Hey Kapil, good call here! I was able to do some composition and was also inspired to take a similar approach using a checker class to do the checking.  There are many more tests now and I have trimmed the test data back as much as I can. For the lambda one, I tried reusing the cross account recording but there is an error in the first describe record that I would need to catch in order for it to work.. so figured I'd just give it a fresh one so I don't need to account for that.